// app/crawler.js
var request = require('request');
var cheerio = require('cheerio');
var debug = require('debug')('appreports:crawler');

var userAgent = 'Mozilla/5.0 (Windows; U; Windows NT 5.1; en-US; rv:1.8.1.9) Gecko/20071025 Firefox/2.0.0.9';

function parseAppName(body) {
  $ = cheerio.load(body);

  var name = $('h1[itemprop="name"]').text();
  return name;
}

exports.getStatus = function(url, callback) {
  request({ url: url, followRedirect: false, headers: { 'User-Agent': userAgent }}, function(err, res) {
    debug(res.statusCode, url);
    return callback(res.statusCode);
  });
}

exports.fetchAppRating = function(url, callback) {
  request({ url: url, followRedirect: false, headers: { 'User-Agent': userAgent }}, function(err, resp, body) {
    if(err) {
      return callback(err);
    }

    $ = cheerio.load(body);

    var ratingCount = $('meta[itemprop="ratingCount"]').attr("content");
    if(!ratingCount) {
      ratingCount = '0';
    }
    ratingCount = ratingCount.replace(',', '');
    ratingCount = parseInt(ratingCount);

    var ratingValue = $('meta[itemprop="ratingValue"]').attr("content");
    ratingValue = ratingValue.replace(',', '.');
    ratingValue = parseFloat(ratingValue);
    return callback(null, isNaN(ratingCount) ? 0 : ratingCount, isNaN(ratingValue) ? 0 : ratingValue);
  });
}

exports.fetchAppDetails = function(url, callback) {
  request({ url: url, followRedirect: false, headers: { 'User-Agent': userAgent }}, function(err, resp, body) {
    if(err) {
      return callback(err);
    }

    var name = parseAppName(body);
    debug('got name', name);
    return callback(null, { name: name });
  });
}

exports.processStoreLinks = function(options, done, progress) {
  if(typeof options === 'function') {
    done = options;
    options = {};
  }

  if(!progress) {
    progress = function(progress) { /* do nothing */ };
  }

  var query = { status: '200' };
  if(options.storeId) {
    query.storeId = options.storeId;
  }

  var stream = StoreLink.find(query).sort({ processedAt: 1 }).limit(options.limit || 0).exec(function(err, storeLinks) {
    var count = 0;
    var max = storeLinks.length;
    debug('processing urls', storeLinks.length);
    var totalNewReviews = 0;
    var max = storeLinks.length;
    var current = 0;
    async.eachLimit(storeLinks, 20, function(storeLink, next) {
      crawler.fetchAppRating(storeLink.url, function(err, ratings, ratingValue) {
        storeLink.processedAt = now().toDate();
        storeLink.save();
        //debug('processing', storeLink.url);

        var data = {
          storeId: storeLink.storeId,
          date: StoreRating.today(),
          region: storeLink.region
        };
        count++;
        (function(data, count, next) {
          StoreRating.findOne(data, data, { upsert: true }, function(err, doc) {
            if(!doc) {
              doc = new StoreRating(data);
            }

            doc.primaryUrl = storeLink.primaryUrl;
            if(doc.ratingCount != ratings) {
              totalNewReviews += ratings - doc.ratingCount || 0;
            }
            doc.ratingCount = ratings;
            doc.ratingAverage = ratingValue;
            doc.ratingTotal = ratings * ratingValue;
            doc.name = storeLink.name;
            doc.platform = storeLink.platform;
            doc.url = storeLink.url;
            doc.bitly = storeLink.bitly;
            doc.baseline = storeLink.baseline;
            debug(doc.region, ratings, ratingValue);
            doc.save();
            current++;
            progress(Math.floor((current / max) * 100));
            return next();
          });
        })(data, count, next);
      });
    }, function() {
      done(totalNewReviews);
    });
  });
}
