// app/date.js
var moment = require('moment-timezone');

function now() {
  return moment().tz('America/Los_Angeles');
}

function nowAsDate() {
  return now().toDate();
}

function todayAsNumber() {
  return now().format('YYYYMMDD') * 1;
}

function todayAsString() {
  return now().format('YYYY-MM-DD');
}

function yesterdayAsNumber() {
  return todayAsNumber() - 1;
}

function getAge(date) {
  return now().diff(moment(date, 'YYYYMMDD'), 'days');
}

function getAgeHuman(date) {
  return moment.duration(now().diff(moment(date, 'YYYYMMDD'), 'days'), 'days').humanize() + ' ago';
}

function convertDateString(date, inFormat, outFormat) {
  if(!date) {
    return null;
  }

  if(!inFormat) {
    inFormat = 'MM/DD/YYYY';
  }

  if(!outFormat) {
    outFormat = 'YYYYMMDD';
  }

  return moment(date, inFormat).format(outFormat);
}

exports.now = now;
exports.nowAsDate = nowAsDate;
exports.todayAsNumber = todayAsNumber;
exports.todayAsString = todayAsString;
exports.yesterdayAsNumber = yesterdayAsNumber;
exports.convertDateString = convertDateString;
exports.getAge = getAge;
exports.getAgeHuman = getAgeHuman;
